src/ui.js

```diff
@@ -78,42 +78,29 @@ function handleFileSelect(map, evt) {
 
     modal.show();
 
-    let i = 0;
     let parseFailures = [];
 
-    let loadFile = (file) => {
-        return new Promise(resolve => {
-            let reader = new FileReader();
-            reader.onload = () => resolve(reader.result);
-            reader.readAsText(file, 'UTF-8');
-        }).then(data => {
-            parseGPX(data, (err, track) => {
-                if (err) {
-                    parseFailures.push({name: file.name, error: err});
-                } else {
-                    track.filename = file.name;
-                    tracks.push(track);
-                }
-
-                modal.progress(++i);
-            });
-        });
-    }
+    let openFile = file => new Promise(resolve => {
+        let reader = new FileReader();
+        reader.onload = () => resolve(reader.result);
+        reader.readAsText(file, "UTF-8");
+    });
 
-    Promise.all(files.map(f => loadFile(f)))
-        .then(() => {
-            tracks.forEach(t => {
-                console.log('Adding track:', t.name);
-                map.addTrack(t);
-            });
+    let loadFile = file => openFile(file).then(parseGPX).then(track => {
+        track.filename = file.name;
+        tracks.push(track);
+        console.info('Adding track:', track.name);
```

Comment:
Unless there's a good reason not to I think I still prefer adding all the tracks to be a separate step in the promise chain. Subjective, but it feels cleaner to me.

Reply:
I was finding that, if I added them all at the end, the progress bar did not make much sense: it would count up without showing anything on the map, then hang at (N-1 of N) or (N of N) while they all got drawn.

Reply:
Good point, I wasn't thinking of it like that!

My initial reason for splitting these steps up was that it seemed significantly faster if it parsed the XML all at once, then did the rendering all at once, rather than doing a full repaint in between each file. A lot of the code has been modified or completely rewritten since then, so it could be that this isn't a valid optimization anymore.

I'll play around with it when I get a chance, but should be fine to merge for now